### PR TITLE
Add JaCoCo to 0.8.8 and upgrade ASM to 9.4

### PR DESCRIPTION
**Pull request by somethingvague**
Related to #16412

Commit 1 of 3

### CODE REVIEW

**Review by cushon — APPROVED** [review]

**Comment by cushon** [comment]
Can you update this to also remove the old jacoco jars, or is there a reason that needs to happen as a follow up?

**Comment by somethingvague** [comment]
> Can you update this to also remove the old jacoco jars, or is there a reason that needs to happen as a follow up?

According to the [README](https://github.com/bazelbuild/bazel/blob/master/third_party/java/jacoco/README.md#3rd-pull-request) the whole change is staged, the removal of the old JARs is the last PR of 3.

Happy to change if that's incorrect?

**Comment by cushon** [comment]
> > Can you update this to also remove the old jacoco jars, or is there a reason that needs to happen as a follow up?
> 
> According to the [README](https://github.com/bazelbuild/bazel/blob/master/third_party/java/jacoco/README.md#3rd-pull-request) the whole change is staged, the removal of the old JARs is the last PR of 3.
> 
> Happy to change if that's incorrect?

Thanks for the context, that sounds good.

File: src/java_tools/junitrunner/java/com/google/testing/coverage/ClassProbesMapper.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public Map<Integer, BranchExp> result() {
   public ClassProbesMapper(String className) {
     classLineToBranchExp = new TreeMap<Integer, BranchExp>();
     stringPool = new StringPool();
-    className = stringPool.get(className);
+    this.className = stringPool.get(className);
```
Comment (cushon) [inline src/java_tools/junitrunner/java/com/google/testing/coverage/ClassProbesMapper.java:55]: I have a separate change out for review for this which should be submitted soon.

**Review by cushon — APPROVED** [review]

**Comment by cushon** [comment]
One last question--did you end up updating to JaCoCo `0.8.8`, or building a later version? Either one is OK with me, but if it's a newer version please mention that in the PR somewhere.

File: src/java_tools/junitrunner/java/com/google/testing/coverage/ClassProbesMapper.java (hunk 1 above)
Comment (cushon) [reply src/java_tools/junitrunner/java/com/google/testing/coverage/ClassProbesMapper.java:55]: You can leave it in for now. The fix should show up in the next hour, and if you want you could rebase at that point. But if not, the merge should be resolved when the PR is imported.

Comment (somethingvague) [reply src/java_tools/junitrunner/java/com/google/testing/coverage/ClassProbesMapper.java]: I've rebased on master and the diff no longer appears for `ClassProbesMapper`

**Comment by kshyanashree** [comment]
Merged at 5030e89ff001a9893a0852a7485ddc8e6510a663

**Pull request closed without merging**